### ISSUE TITLE
Integrating Ideas form FTPipe/Mixed-pipe

### ISSUE THREAD

**Issue by saareliad** [pytorch/PiPPy#236]
Hi, nice repo,
I'd like to check how our work can contribute, please checkout

- paper  https://www.usenix.org/system/files/atc21-eliad.pdf
- code https://github.com/saareliad/FTPipe

The following was used mostly for research a couple of years ago (started early ~2019, ended ~2021), but I think that a framework supported by PyTorch is the real way to go for the broad usage of pipeline parallelism.
Anyway, I was waiting for Pytorch to implement something like this for a while! 

While our paper discusses mostly fine-tuning of giant DNNs, which at the time were identified as the main use-cases for Pipeline Parallelism with commodity HW  (- i.e., memory/communication constrained GPUs with excellent perf/cost which are commonly used, vs. high end GPUs and massive clusters typically used for large-scale pertaining). I think there are several more ideas we explored and implemented which can be used, at least conceptually.

**Partitioning/Splitting**: combining and comparing several splitting methods, with various granularity levels, memory awareness, and so on. Especially, handling encoder-decoders like T5 (See Mixed-Pipe), and providing tradeoffs between fine-granularity, partitioning speed, training speed, and so on. 
Note that an "optimal split" may not be the practical way to go - since you have to consider the granularity of layers/ops and so on. So what seems optimal under coarse granularity is not really optimal, and there may be a need to play with granularity levels for certain models.

**Training/partitioning Recipes for model type and size** e.g., giant transformers. I think it is especially interesting since we empirically found the accuracy barely goes down if not at all when using the simplest asynchronous pipeline for giant DNNs and there is a great speed up for doing so, even when artificially generating more stages than GPUs.
(However, we did encounter an accuracy drop for asynchronous pipelines with smaller DNNs like resent/WRN/ViT-base/small Transformers like  distilled-bert .\
 While many small models typically do not need any Pipeline-parallelism and can work well with data parallelism, for certain models - using a synchronous pipeline or even an async pipeline with advanced scheduling and staleness mitigation is very relevant. )

**More Scheduling techniques and combinations, advanced async staleness mitigation**: we have explored many pipeline scheduling methods and their tradeoffs - not all covered in the paper ( sync, async, async with _staeleness mitigation_ techniques, combined with necessary adaptations to optimizers, including gradient accumulation, recomputation, and so on - which are all pluggable.)
(later, there are some more advanced and low-level things there like integrating staleness mitigation with the  LRscheduler, adapting the optimizer math, generating the mathematical corrections for momentum automatically with symbolic computation, and so on ). 
Anyway, If going async these techniques generally improve accuracy, adding some compute-memory-accuracy tradeoffs. 
I'd like to see people research this as well - we have collaborated with a large company that implemented some of these ideas for training key models and they reported success.

There are also other important concepts for PP we explored (e.g., tracing, profiling)
**Large model Profiling** - well profiling is basic - But we implemented methods to do so and handle giant DNNs while saving memory. I think that lazy loading was introduced and it can be combined. Anyway, considering giant DNNs,  using "synthetic profiling" of backdrop saves tones of memory works fast and well. 

**Tracing**: We had implemented a jit. trace-based pipeline in the past (~2018-2019 way before fx) back then it was problematic to trace NLP models so we replaced it and we used a python tracer, which can also catch other computations. Not sure if this is still relevant since both jit and fx improved greatly and time has passed since - but you can consider also combining such tracing for more robust and automatic usage.
Furthermore, we used the names generated by tracing for controlling pipeline partitioning granularity when needed.

I think this repo will advance research and hope to use it myself soon!
Best of luck
